task_processing/plugins/mesos/execution_framework.py

```diff
@@ -15,338 +13,118 @@
     def __init__(
         self,
-        name,
-        staging_timeout,
-        reconcile_backoff=1,
-        reconcile_start_time=float('inf'),
-        dry_run=False,
+        name="remote_run",
+        task_staging_timeout_s=240,
+        pool=None,
+        max_task_queue_size=1000,
         translator=mesos_status_to_event,
+        slave_blacklist_timeout_s=900,
+        offer_backoff=240,
+        task_retries=2,
     ):
-        self.translator = translator
-        self.queue = Queue(1000)
         self.name = name
-        self.tasks_with_flags = {}
-        self.constraint_state = {}
-        self.constraint_state_lock = threading.Lock()
-        self.frozen = False
+        # wait this long for a task to launch.
+        self.task_staging_timeout_s = task_staging_timeout_s
         self.framework_info = self.build_framework_info()
-        self.tasks_waiting = Queue(1000)
-        self.dry_run = dry_run
-
-        # don't accept resources until we reconcile.
-        self.reconcile_start_time = reconcile_start_time
+        self.pool = pool
+        self.translator = translator
+        self.slave_blacklist_timeout_s = slave_blacklist_timeout_s
+        self.offer_backoff = offer_backoff
+        self.task_retries = task_retries
 
-        # wait this long after starting a reconcile before accepting offers.
-        self.reconcile_backoff = 1
+        self.task_queue = Queue(max_task_queue_size)
+        self.task_update_queue = Queue(max_task_queue_size)
+        self.driver = None
+        self.are_offers_suppressed = False
 
-        # wait this long for a task to launch.
-        self.staging_timeout = staging_timeout
+        self.offer_decline_filter = self.build_decline_offer_filter()
+        # TODO: These should be thread safe
+        self.blacklisted_slaves = {}
+        self.task_metadata = {}
 
-        # Gets set when registered() is called
-        self.framework_id = None
+        self.start_new_thread(self.kill_tasks_stuck_in_staging)
+        self.start_new_thread(self.unblacklist_slaves)
 
-        self.blacklisted_slaves = set()
-        self.blacklisted_slaves_lock = threading.Lock()
-        self.blacklist_timeout = 10
+    def start_new_thread(self, func):
+        t = threading.Thread(target=func, args=())
+        # t.set_daemon = True
+        t.start()
 
-    def shutdown(self, driver):
-        # TODO: this is naive, as it does nothing to stop on-going calls
-        #       to statusUpdate or resourceOffers.
-        paasta_print(
-            'Freezing the scheduler. Further status updates and resource '
-            'offers are ignored.'
-        )
-        self.frozen = True
-        paasta_print("Killing any remaining live tasks.")
-        for task, parameters in self.tasks_with_flags.items():
-            if parameters.mesos_task_state in LIVE_TASK_STATES:
-                self.kill_task(driver, task)
+    def kill_tasks_stuck_in_staging(self):
```

Comment:
I was thinking if we launch 1000 containers, that would mean we will spawn 1000 new threads.  Since the task state is stored in a single datastructure, the threads will have to acquire locks(going to add a mutex for all the shared datastructures). Seems like an unnecessary overhead. We can achieve the same thing with a single thread and up to 10 sec delay.